enhancements/cluster-capabilities-api.md

```diff
@@ -0,0 +1,328 @@
+`capabilities.config.openshift.io`, is created to hold details about
+the configured features of a cluster. It will include two fields to
+start, and may be expanded over time.
+
+The `highAvailabilityMode` field will use an enum to express the
+high-availability expectations. Possible values are `full` and
+`none`. The default is `full`, which represents the behavior operators
+have today in a "normal" cluster. The `none` mode will be used in
+single-node deployments (developer and production), and the operator
+should not configure the operand for highly-available
+operation. Systems such as the assisted installer, which may
+experience periods early in the lifetime of the cluster when full HA
+is not possible, can start with `none` and transition to `full` when
+the cluster is more fully online.
+
+The `resourceLimits` fields will use an enum to express the general
+expectations for how much capacity operators should give their
+operands. Valid values for the enum are `small`, `medium`, and
+`large`. The default is `large`, representing the behavior for
+operators today. The `medium` configuration would be used to lower the
+resource use in single-node production clusters. For example, the
+cluster-monitoring-operator might deploy prometheus with a reduced set
+of rules and scraping targets along with a shorter retention period,
+and it might not deploy grafana at all. The `small` configuration
+would be used to lower the resource use even further, and might be
+used in a single-node developer cluster.
+
+The new cluster-wide `capabilities` API will supersede existing
+unsupported APIs, such as the flags to disable high-availability
+requirements for cluster-etcd-operator and
+cluster-authentication-operator.
+
+If other *supported* APIs for configuring resource use exist, they
+take precedence over this new cluster-wide `capabilities` API. The
+cluster-wide API should be treated as a default for cases where
+operator-specific values are not set, a replacement for unsupported
+APIs, and a way to avoid adding new operator-specific APIs.
+
+### User Stories
+
+1. As an author of an operator, I want to make my operator adapt to
+   clusters with different sizes and topologies so that more users can
+   benefit from using the operator.
+2. As the designer of a cluster profile, I want to express the intent
+   of the profile so that operators can adapt to the new cluster type.
+
+### Implementation Details/Notes/Constraints
+
+We will add the new `capabilities` API struct, enums, and CRD to the
+`openshift/api` repository, following the pattern used by the `dns`
+API.
+
+We will add logic to the installer to use the profile name to create a
+manifest to set up the singleton of the new `capabilities` API
+resource.
+
+### Risks and Mitigations
+
+What are the risks of this proposal and how do we mitigate. Think broadly. For
+example, consider both security and how this will impact the larger OKD
+ecosystem.
```

Comment:
this section still needs filling in.  I think the main risk is that we flub the initial schema and end up providing an API that no components actually need.  Might be worth compiling a list of the existing config properties we're replacing (just the etcd and auth HA flags?) and the per-component configs we are avoiding (some monitoring knob?  Anything else?).  The conservative approach would be to leave these knobs per-component until we reach some critical mass to make a stronger "look at all these per-component knobs we can quietly de-emphasize by growing this cluster-scoped knob" case.

Reply:
The driver for this requirement is the need to reduce the overall footprint of the cluster, rather than any specific component, right?

How would the installer set the component-specific API fields?

Reply:
I've added details, but I honestly don't see a way to do this with operator-specific APIs that doesn't involve a huge amount of work to tie the installer to those APIs so they can be set during deployment.